### DISCUSSION TITLE
Invalid WebSocket upgrade request

### DISCUSSION THREAD

**Discussion by yashnb-serverpod** [serverpod/serverpod#1122] [Q&A] [answered]
I have hosted serverpod container on local office server, it is running perfectly for API req/responses but when i try to open streaming connection by calling **client.openStreamingConnection()**, it is giving error on server log as follows :

**2023-07-11 13:05:49.446755Z Internal server error. Zoned exception.
WebSocketException: Invalid WebSocket upgrade request**

anyone can help me in such case?
@vlidholt

**Comment by vlidholt** [comment]
It's a little bit hard to guess what is going on here without seeing more code and understand what you are doing. Can you please elaborate a bit?

**Reply by yashnb-serverpod** [reply]
sure, Thank you very much for your quick response.

We have moved whole serverpod project dir containing proj_server, proj_client, proj_flutter on server and created postgres instance.

ran the server as follows

**docker-compose up --build --detach** in proj_server dir
**dart bin/main.dart** in proj_server dir

after that we have running server serving on 8080

now in server_proj we have created one endpoint which has 3 overridden methods 
**streamOpened
streamClosed
handleStreamMessage**


in my flutter project on very first screen after login inside initState we are calling 
**client.openStreamingConnection()**
which is not opening socket and we are getting error on server and flutter both logs

**flutter :  unable to connect to baseUrl : 0/websocket/#
server : WebSocketException: Invalid WebSocket upgrade request**

**Reply by yashnb-serverpod** [reply]
@vlidholt it's bit urgent for me, is there any particular cause for this? as I can see it is exception thrown when dart transform connection called for websocket, and in many dart projects there are similar issues open, if possible is there any working example of **openStreamingConnection**  in any project, please share server and flutter app side code , so that I can verify..

**Reply by vlidholt** [reply]
https://pixorama.live is a full example of streaming connections with Serverpod (demo and source code).

- Are you using the same version of Serverpod in all packages (client/server/flutter)? You can validate by removing the  `^` in your pubspec file.
- Have you ran `serverpod generate`?
- Are you using the same version of Serverpod's CLI as the serverpod packages you are using (check with `serverpod version`
- Does it work when you run Serverpod locally?

As a side note, `openStreaminConnection` should work, but you may want to use the `StreamingConnectionHandler` as it will provide a nicer interface for reliably reconnect to your server.

**Reply by yashnb-serverpod** [reply]
thanks for your reply.

yes versions are same and I have ran generate
it works locally too.
am trying with StreamingConnectionHandler and checking if I can get more info, else will ping you as am running out of time...

**Reply by yashnb-serverpod** [reply]
@vlidholt I have added **StreamingConnectionHandler** and following are the logs. it is also giving same exception

I/flutter (10952): waitingToRetry
I/flutter (10952): received new streaming state : connecting
E/flutter (10952): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: WebSocketException: Connection to 'http://st-serverpod.in:0/websocket#' was not upgraded to websocket
E/flutter (10952): #0      _WebSocketImpl.connect (dart:_http/websocket_impl.dart:1011:41)
E/flutter (10952): #1      WebSocket.connect (dart:_http/websocket.dart:320:22)
E/flutter (10952): #2      new IOWebSocketChannel.connect (package:web_socket_channel/io.dart:81:28)
E/flutter (10952): #3      connect (package:web_socket_channel/src/_connect_io.dart:15:24)
E/flutter (10952): #4      new WebSocketChannel.connect (package:web_socket_channel/src/channel.dart:115:16)
E/flutter (10952): #5      ServerpodClientShared.openStreamingConnection (package:serverpod_client/src/serverpod_client_shared.dart:226:37)
E/flutter (10952): <asynchronous suspension>
E/flutter (10952): 
I/flutter (10952): received new streaming state : waitingToRetry
I/flutter (10952): received new streaming state : waitingToRetry
I/flutter (10952): received new streaming state : waitingToRetry
I/flutter (10952): received new streaming state : waitingToRetry
I/flutter (10952): received new streaming state : waitingToRetry
I/flutter (10952): received new streaming state : connecting
E/flutter (10952): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: WebSocketException: Connection to 'http://st-serverpod.in:0/websocket#' was not upgraded to websocket
E/flutter (10952): #0      _WebSocketImpl.connect (dart:_http/websocket_impl.dart:1011:41)
E/flutter (10952): #1      WebSocket.connect (dart:_http/websocket.dart:320:22)
E/flutter (10952): #2      new IOWebSocketChannel.connect (package:web_socket_channel/io.dart:81:28)
E/flutter (10952): #3      connect (package:web_socket_channel/src/_connect_io.dart:15:24)
E/flutter (10952): #4      new WebSocketChannel.connect (package:web_socket_channel/src/channel.dart:115:16)
E/flutter (10952): #5      ServerpodClientShared.openStreamingConnection (package:serverpod_client/src/serverpod_client_shared.dart:226:37)
E/flutter (10952): <asynchronous suspension>
E/flutter (10952): 
I/flutter (10952): received new streaming state : waitingToRetry
I/flutter (10952): received new streaming state : waitingToRetry
I/flutter (10952): received new streaming state : waitingToRetry
I/flutter (10952): received new streaming state : waitingToRetry
I/flutter (10952): received new streaming state : waitingToRetry

**Reply by vlidholt** [reply, marked as answer]
If it works locally, it's probably not an issue with your code, but with the setup of your server machine and perhaps its firewall.

**Reply by yashnb-serverpod** [reply]
got it, I have asked DevOps for the same, trying to fix it, otherwise will move to aws as it's documented on serverpod.

**Reply by mkoehne** [reply]
@yashnb-serverpod We have the same problem with Nginx run on our server. Did you find a solution for the problem?

**Reply by yahu1031** [reply]
Most likely the `user-agent`headers. Can you guys cross check them. The dart will send `Dart/3.0 (dart:io)` as default headers.
